framework/src/userobject/ElementQualityChecker.C

```diff
@@ -0,0 +1,201 @@
+/* MOOSE - Multiphysics Object Oriented Simulation Environment  */
+/*                                                              */
+/*          All contents are licensed under LGPL V2.1           */
+/*             See LICENSE for full restrictions                */
+/****************************************************************/
+#include "ElementQualityChecker.h"
+#include "MooseError.h"
+
+#include "libmesh/elem_quality.h"
+#include "libmesh/enum_elem_quality.h"
+
+MooseEnum
+ElementQualityChecker::QualityMetricType()
+{
+  return MooseEnum("ASPECT_RATIO SKEW SHEAR SHAPE MAX_ANGLE MIN_ANGLE CONDITION DISTORTION TAPER "
+                   "WARP STRETCH DIAGONAL ASPECT_RATIO_BETA ASPECT_RATIO_GAMMA SIZE JACOBIAN");
+}
+
+MooseEnum
+ElementQualityChecker::FailureMessageType()
+{
+  return MooseEnum("WARNING ERROR", "WARNING");
+}
+
+template <>
+InputParameters
+validParams<ElementQualityChecker>()
+{
+  InputParameters params = validParams<ElementUserObject>();
+  params.addClassDescription("Class to check the quality of each element using different metrics "
+                             "from libmesh.");
+
+  params.addRequiredParam<MooseEnum>("metric_type",
+                                     ElementQualityChecker::QualityMetricType(),
+                                     "Type of quality metric to be checked");
+  params.addParam<Real>("upper_bound", "the upper bound for provided metric type");
+  params.addParam<Real>("lower_bound", "The lower bound for provided metric type");
+  params.addParam<MooseEnum>("failure_type",
+                             ElementQualityChecker::FailureMessageType(),
+                             "The way how the failure of quality metric check should respond");
+  params.set<MultiMooseEnum>("execute_on") = "initial";
+
+  return params;
+}
+
+ElementQualityChecker::ElementQualityChecker(const InputParameters & parameters)
+  : ElementUserObject(parameters),
+    _m_type(getParam<MooseEnum>("metric_type").getEnum<ElemQuality>()),
+    _has_upper_bound(isParamValid("upper_bound")),
+    _has_lower_bound(isParamValid("lower_bound")),
+    _upper_bound(_has_upper_bound ? getParam<Real>("upper_bound") : 0.0),
+    _lower_bound(_has_lower_bound ? getParam<Real>("lower_bound") : 0.0),
+    _failure_type(getParam<MooseEnum>("failure_type").getEnum<FailureType>())
+{
+  if (_has_lower_bound && _has_upper_bound && _lower_bound >= _upper_bound)
+    mooseError("Provided lower bound should be less than provided upper bound!");
+}
+
+void
+ElementQualityChecker::initialize()
+{
+  _m_values.clear();
+  _elem_ids.clear();
+  _bypassed = 0;
+}
+
+void
+ElementQualityChecker::execute()
+{
+  // obtain the available quality metric for current ElemType
+  std::vector<ElemQuality> metrics_avail = libMesh::Quality::valid(_current_elem->type());
+
+  // check whether the provided quality metric is applicable to current ElemType
+  if (!checkMetricApplicability(_m_type, metrics_avail))
+  {
+    _bypassed = 1;
```

Comment:
You only have one value for bypassed but you could have mixed types where some are applicable and some are not. You don't know when visiting which order you will hit these elements in. It's possible to hit elements that you do want to check after elements that are bypassed. This logic is faulty in that case.

Reply:
I'm still confused as to what _bypassed is supposed to do. You are setting it to 1 here, but you are summing on it in threadJoin and finalize. That means it's dependent on the number of processors you are running on. That can't be right.